# Sentiment Analysis with Amazon Comprehend
                                       -- In Application with South Park Review 

In this notebook, I used Amazon comprehend to implement sentiment analysis on one review of the episode of South Park from Dani Di Placido Forbes **#OnTV** section (published Nov. 9, 2017). 

![](http://imagecurl.com/images/04065054009291592566_thumb.jpg)

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Plot-Excerpt" data-toc-modified-id="Plot-Excerpt-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Plot Excerpt</a></span></li><li><span><a href="#Setting-up-Environment" data-toc-modified-id="Setting-up-Environment-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Setting up Environment</a></span></li><li><span><a href="#Sentiment-Analysis-with-Amazon-Comprehend" data-toc-modified-id="Sentiment-Analysis-with-Amazon-Comprehend-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Sentiment Analysis with Amazon Comprehend</a></span><ul class="toc-item"><li><span><a href="#On-a-text-line" data-toc-modified-id="On-a-text-line-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>On a text line</a></span></li><li><span><a href="#On-a-line-in-review-document" data-toc-modified-id="On-a-line-in-review-document-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>On a line in review document</a></span></li><li><span><a href="#On-an-entire-movie-review-document" data-toc-modified-id="On-an-entire-movie-review-document-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>On an entire movie review document</a></span></li></ul></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Conclusion</a></span></li><li><span><a href="#References" data-toc-modified-id="References-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>References</a></span></li></ul></div>

## Plot Excerpt

- First thing first, let's look at the plot excerpt of this episode 'Doubling Down': 

Cartman's abusive treatment of Heidi Turner is straining their relationship. He blames his anger issues on his poor diet, so Heidi convinces Cartman to try a vegan diet with her in an attempt to reconcile. The next day at lunch, Cartman continues to degrade Heidi to the rest of the boys, but treats her kindly when she shows up. When Kyle talks to Heidi after school about her relationship with Cartman, she gets defensive and avoids the subject. At Heidi's home, Cartman states that he now enjoys Beyond Meat and presents her with a bucket of KFC to try, claiming that it is actually "Beyond KFC" and vegan. Back at school, she feels bloated and ill afterwards while Cartman insults Heidi behind her back and tells the boys that she is getting fat. Kyle goes to a girls' volleyball practice to ask them to stop mocking Heidi, and they question if Kyle is doing so because he has feelings for Heidi himself. He decides to back off his attempts to break up Cartman and Heidi, but later talks to Heidi in the school gymnasium. The two bond as Kyle tells Heidi that Cartman is an expert at playing the victim and will never change.

## Setting up Environment

In AWS account:
- create a [new IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html#id_users_create_cliwpsapi)
- note down 1) **YOUR_ACCESS_KEY** 2) **YOUR_SECRET_KEY** once the IAM user sets up

In terminal:
- pip install awscli
- aws configure
  - aws_access_key_id = **YOUR_ACCESS_KEY**
  - aws_secret_access_key = **YOUR_SECRET_KEY**
  - Default region name = **us-east-1**
  - Default output format = **json**

In [1]:
import pandas as pd
import boto3
import json

## Sentiment Analysis with Amazon Comprehend 

### On a text line 

In [2]:
comprehend = boto3.client(service_name='comprehend', region_name='us-east-1')

In [3]:
text = "Garrison Trump’s solution is to violently rape all three, as he nurses a black eye and defends Trump’s actions, is painfully accurate. The parallel between Heidi and Cartman’s poisonous, yet stubborn relationship is extended to Trump’s supporters, who watch Trump’s latest gaffe on tv, and look away, unwilling to criticize him for fear of the left’s rabid gloating. This is unfortunately, exactly what’s happening." 

print('Calling DetectSentiment')
print(json.dumps(comprehend.detect_sentiment(Text=text, LanguageCode='en'), sort_keys=True, indent=4))
print('End of DetectSentiment\n')

Calling DetectSentiment
{
    "ResponseMetadata": {
        "HTTPHeaders": {
            "connection": "keep-alive",
            "content-length": "163",
            "content-type": "application/x-amz-json-1.1",
            "date": "Tue, 13 Mar 2018 03:18:50 GMT",
            "x-amzn-requestid": "3ff2887f-266d-11e8-b48a-9f02d80b9bc8"
        },
        "HTTPStatusCode": 200,
        "RequestId": "3ff2887f-266d-11e8-b48a-9f02d80b9bc8",
        "RetryAttempts": 0
    },
    "Sentiment": "NEGATIVE",
    "SentimentScore": {
        "Mixed": 0.04560389369726181,
        "Negative": 0.5313727259635925,
        "Neutral": 0.41348445415496826,
        "Positive": 0.009538856334984303
    }
}
End of DetectSentiment



**Seems that the review for Garrison Trump in this episode is mostly negative (a dominant proportion of 53%), and somewhat neutral (41%).**

**The result is not surprising, as in this episode President Garrison keeps on his campaign promise to "fxxk them all to death", and threaten those who against him.** 

![Garrison](https://hips.hearstapps.com/esq.h-cdn.co/assets/16/42/980x544/gallery-1476975335-south-park.png?resize=768:*)

### On a line in review document

In [4]:
path = "~/Natural_Language_Processing/SouthParkReview.txt"
doc1 = open(path, "r")
output = doc1.readlines()

     Author Note: there is a text size limit for Detectsemtiment analysis in AWS of 5000 bytes. So I have to manually cut down the text file and make the size fit.

In [5]:
output[1]

'The episode focuses on my favorite subplot of the season, Heidi and Cartman’s relationship. \n'

In [6]:
print(json.dumps(comprehend.detect_sentiment(Text=output[1], LanguageCode='en'), sort_keys=True, indent=4))

{
    "ResponseMetadata": {
        "HTTPHeaders": {
            "connection": "keep-alive",
            "content-length": "163",
            "content-type": "application/x-amz-json-1.1",
            "date": "Tue, 13 Mar 2018 03:18:50 GMT",
            "x-amzn-requestid": "400d631f-266d-11e8-9d52-a71b4437c023"
        },
        "HTTPStatusCode": 200,
        "RequestId": "400d631f-266d-11e8-9d52-a71b4437c023",
        "RetryAttempts": 0
    },
    "Sentiment": "NEUTRAL",
    "SentimentScore": {
        "Mixed": 0.0037473251577466726,
        "Negative": 0.008039911277592182,
        "Neutral": 0.6919484734535217,
        "Positive": 0.2962643802165985
    }
}


**The review for this episode is neutral (a dominant proportion of 69%) to positive (30%). Key word 'favorite' contributes mostly to the positive sentiment.**

![Heidi and Cartman](https://i.ytimg.com/vi/EhF0VR1PFRE/maxresdefault.jpg)

### On an entire movie review document

In [7]:
whole_doc = ', '.join(map(str, output))
print(json.dumps(comprehend.detect_sentiment(Text=whole_doc, LanguageCode='en'), sort_keys=True, indent=4))

{
    "ResponseMetadata": {
        "HTTPHeaders": {
            "connection": "keep-alive",
            "content-length": "161",
            "content-type": "application/x-amz-json-1.1",
            "date": "Tue, 13 Mar 2018 03:18:50 GMT",
            "x-amzn-requestid": "402642ce-266d-11e8-a22e-41f36acf1541"
        },
        "HTTPStatusCode": 200,
        "RequestId": "402642ce-266d-11e8-a22e-41f36acf1541",
        "RetryAttempts": 0
    },
    "Sentiment": "NEUTRAL",
    "SentimentScore": {
        "Mixed": 0.08717764168977737,
        "Negative": 0.13139696419239044,
        "Neutral": 0.6841796636581421,
        "Positive": 0.09724566340446472
    }
}


## Conclusion

Regardless of an negative attitude towards Garrison, the author's attitude towards the episode "Doubling down" of South Park is netural (score: 0.68) to somewhat negative (score: 0.13). 

## References

- ['South Park' Review: 'Doubling Down' Is The Most Insightful Episode In Years](https://www.forbes.com/sites/danidiplacido/2017/11/09/south-park-review-doubling-down-is-the-most-insightful-episode-in-years/#171864e17684)
- <a href="https://en.wikipedia.org/wiki/Doubling_Down_(South_Park)#Plot">'South Park' Season 21 Episode 7 'Doubling Down'</a>